### PR TITLE
JAMES-3856 RFC-9208 Implement IMAP QUOTA revision

### PR DESCRIPTION
**Pull request by chibenwa**
- Adds QUOTA=RES-MESSAGE and QUOTA=RES-STORAGE capabilities
 - Adds support for DELETED status item (also mandated by RFC-9051 IMAP4Rev2)
 - Adds support for DELETED_STORAGE status item
 - Adds the optional OVERQUOTA response code for append commands
 - Document support for this specification

### CODE REVIEW

**Comment by quantranhong1999** [comment]
`org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:check (check-style) on project protocols-imap: You have 1 Checkstyle violation.`

File: protocols/imap/src/main/java/org/apache/james/imap/api/ImapConstants.java [hunk 1]
```diff
@@ -152,6 +154,10 @@ public interface ImapConstants {
 
     String STATUS_SIZE = "SIZE";
 
+    String STATUS_DELETED = "DELETED";
+
+    String STATUS_DELETED_STORAGE = "DELETED_STORAGE";
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/api/ImapConstants.java:159]:
```suggestion
    String STATUS_DELETED_STORAGE = "DELETED-STORAGE";
```

`The DELETED and DELETED-STORAGE status data items allow for estimation of the amount of resources that could be freed by an EXPUNGE on a mailbox.`

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java [hunk 2]
```diff
@@ -30,10 +30,15 @@ public enum StatusItem {
...
         UID_VALIDITY,
         UNSEEN,
-        HIGHEST_MODSEQ
+        HIGHEST_MODSEQ,
+        // See https://www.iana.org/go/rfc8438
+        SIZE,
+        // See https://www.rfc-editor.org/rfc/rfc9208.html
+        DELETED,
+        // See https://www.rfc-editor.org/rfc/rfc9208.html
+        DELETED_STORAGE
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java:41]:
```suggestion
        DELETED-STORAGE
```

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java [hunk 3]
```diff
@@ -70,6 +75,14 @@ public boolean isSize() {
         return statusItems.contains(StatusItem.SIZE);
     }
 
+    public boolean isDeleted() {
+        return statusItems.contains(StatusItem.DELETED);
+    }
+
+    public boolean isDeletedStorage() {
+        return statusItems.contains(StatusItem.DELETED_STORAGE);
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java:83]:
```suggestion
        return statusItems.contains(StatusItem.DELETED-STORAGE);
```

File: protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java [hunk 4]
```diff
@@ -182,6 +185,29 @@ private StatusDataItems.StatusItem readMessages(ImapRequestLineReader request) t
...
+        if (c == '_') {
+            assertChar(request, '_', '_');
+            assertChar(request, 's', 'S');
+            assertChar(request, 't', 'T');
+            assertChar(request, 'o', 'O');
+            assertChar(request, 'r', 'R');
+            assertChar(request, 'a', 'A');
+            assertChar(request, 'g', 'G');
+            assertChar(request, 'e', 'E');
+            return StatusDataItems.StatusItem.DELETED_STORAGE;
```
Comment (quantranhong1999) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java:206]:
```suggestion
            return StatusDataItems.StatusItem.DELETED-STORAGE;
```

File: protocols/imap/src/test/java/org/apache/james/imap/encode/MailboxStatusResponseEncoderTest.java [hunk 5]
```diff
@@ -53,10 +53,13 @@ void testDoEncode() throws Exception {
...
+        final Long size = 42L;
+        final Long deleted = 23L;
+        final Long deletedStorage = 13L;
         final String mailbox = "A mailbox named desire";
 
-        encoder.encode(new MailboxStatusResponse(null, messages, recent, uidNext,
+        encoder.encode(new MailboxStatusResponse(null, null, deletedStorage, messages, recent, uidNext,
                 null, uidValidity, unseen, mailbox), composer);
-        assertThat(writer.getString()).isEqualTo("* STATUS \"A mailbox named desire\" (MESSAGES 2 RECENT 3 UIDNEXT 5 UIDVALIDITY 7 UNSEEN 11)\r\n");
+        assertThat(writer.getString()).isEqualTo("* STATUS \"A mailbox named desire\" (MESSAGES 2 DELETED_STORAGE 13 RECENT 3 UIDNEXT 5 UIDVALIDITY 7 UNSEEN 11)\r\n");
```
Comment (quantranhong1999) [inline protocols/imap/src/test/java/org/apache/james/imap/encode/MailboxStatusResponseEncoderTest.java:63]:
```suggestion
        assertThat(writer.getString()).isEqualTo("* STATUS \"A mailbox named desire\" (MESSAGES 2 DELETED-STORAGE 13 RECENT 3 UIDNEXT 5 UIDVALIDITY 7 UNSEEN 11)\r\n");
```

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java (hunk 2 above)
Comment (Arsnael) [reply protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java:41]: Disagree, this is a constant name.

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java (hunk 3 above)
Comment (Arsnael) [reply protocols/imap/src/main/java/org/apache/james/imap/api/message/StatusDataItems.java:83]: Disagree too

File: protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java (hunk 4 above)
Comment (Arsnael) [reply protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java:206]: Disagree too

File: protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java [hunk 6]
```diff
@@ -182,6 +185,29 @@ private StatusDataItems.StatusItem readMessages(ImapRequestLineReader request) t
...
+        assertChar(request, 'd', 'D');
+        assertChar(request, 'e', 'E');
+        assertChar(request, 'l', 'L');
+        assertChar(request, 'e', 'E');
+        assertChar(request, 't', 'T');
+        assertChar(request, 'e', 'E');
+        assertChar(request, 'd', 'D');
+        char c = request.nextWordChar();
+        if (c == '_') {
+            assertChar(request, '_', '_');
```
Comment (Arsnael) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/parser/StatusCommandParser.java:198]: should be '-' instead of '_' (line above as well)

**Review by Arsnael — APPROVED** [review]

**Pull request merged**